File: src/langnostic/scratch.clj
(ns langnostic.scratch
  (:require [clojure.java.io :as io]
            [clojure.string :as str]
            [cheshire.core :as json]))

(defn next-id []
  (-> "posts.json" io/resource io/reader line-seq last json/decode (get "id") inc))

(defn all-tags []
  (->> "posts.json" io/resource io/reader line-seq
       (map #(json/parse-string % (fn [k] (keyword (.toLowerCase k)))))
       (mapcat :tags) frequencies
       (into (sorted-map))))

(defn slug-from-file [file] (str/replace (.getName file) #".md$" ""))

(defn title-from-file [file]
  (->> (str/split (slug-from-file file) #"-")
       (map (fn [word]
              (if (#{"and" "of"} word)
                word
                (str/capitalize word))))
       (str/join " ")))

(defn drafts-by-prefix [prefix]
  (->> (io/file "drafts")
       file-seq
       (filter
        (fn [f]
          (and
           (str/starts-with? (.getName f) prefix)
           (not (str/ends-with? (.getName f) "~")))))))

(defn new [file-prefix & {:keys [title tags] :or {tags []}}]
  (let [drafts (drafts-by-prefix file-prefix)]
    (if (= 1 (count drafts))
      (let [file (first drafts)
            slug (slug-from-file file)
            title (or title (title-from-file file))
            record {:id (next-id) :title title
                    :file (slug-from-file file)
                    :posted (quot (System/currentTimeMillis) 1000)
                    :tags tags}]
        (io/copy file (io/file (io/resource "posts") (.getName file)))
        (io/delete-file file)
        (spit (io/file (io/resource "posts.json")) (str (json/encode record) "\n") :append true))
      (println "MULTIPLE DRAFTS FOUND" drafts))))

File: src/langnostic/scratch.py
import json
import os
import re
import shutil
import time
from collections import Counter
from pathlib import Path
from typing import Dict, Iterable, List, Optional, Union

RESOURCES_DIR = Path("resources")
POSTS_JSON = RESOURCES_DIR / "posts.json"
POSTS_DIR = RESOURCES_DIR / "posts"
DRAFTS_DIR = Path("drafts")


Pathish = Union[str, os.PathLike]


def _read_json_lines(path: Path = POSTS_JSON) -> List[Dict]:
    with path.open("r", encoding="utf-8") as f:
        return [json.loads(line) for line in f if line.strip()]


def next_id() -> int:
    """Return one greater than the id of the last post in resources/posts.json."""
    records = _read_json_lines(POSTS_JSON)
    return int(records[-1]["id"]) + 1


def all_tags() -> Dict[str, int]:
    """Return tag frequencies from resources/posts.json, sorted by tag."""
    tag_counts = Counter()

    for raw in _read_json_lines(POSTS_JSON):
        post = {k.lower(): v for k, v in raw.items()}
        tag_counts.update(post.get("tags", []))

    return dict(sorted(tag_counts.items()))


def slug_from_file(file: Pathish) -> str:
    """Return the filename without a trailing .md extension."""
    name = Path(file).name
    return re.sub(r"\.md$", "", name)


def title_from_file(file: Pathish) -> str:
    """Generate a title from a markdown filename slug."""
    small_words = {"and", "of"}

    words = []
    for word in slug_from_file(file).split("-"):
        if word in small_words:
            words.append(word)
        else:
            words.append(word.capitalize())

    return " ".join(words)


def drafts_by_prefix(prefix: str) -> List[Path]:
    """
    Return draft files whose filename starts with prefix,
    excluding editor backup files ending in ~.
    """
    if not DRAFTS_DIR.exists():
        return []

    return [
        path
        for path in DRAFTS_DIR.rglob("*")
        if path.is_file()
        and path.name.startswith(prefix)
        and not path.name.endswith("~")
    ]


def new(
    file_prefix: str,
    *,
    title: Optional[str] = None,
    tags: Optional[Iterable[str]] = None,
) -> Optional[Dict]:
    """
    Publish exactly one matching draft.

    Copies the draft into resources/posts/, deletes the draft, and appends a
    JSON-lines record to resources/posts.json.

    Example:
        scratch.new("my-post-prefix", tags=["python", "blog"])
        scratch.new("my-post-prefix", title="Custom Title", tags=["meta"])
    """
    drafts = drafts_by_prefix(file_prefix)

    if len(drafts) != 1:
        print("MULTIPLE DRAFTS FOUND", drafts)
        return None

    file = drafts[0]
    slug = slug_from_file(file)
    post_title = title or title_from_file(file)

    record = {
        "id": next_id(),
        "title": post_title,
        "file": slug,
        "posted": int(time.time()),
        "tags": list(tags or []),
    }

    POSTS_DIR.mkdir(parents=True, exist_ok=True)

    destination = POSTS_DIR / file.name
    shutil.copy2(file, destination)
    file.unlink()

    with POSTS_JSON.open("a", encoding="utf-8") as f:
        f.write(json.dumps(record, separators=(",", ":")) + "\n")

    return record
